Add CockroachDB module

leourbina opened the pull request:
# Adds CockroachDB Support

This is mainly the same as the postgres container, with a few minor details:

* It uses CRDB's `--insecure` flag to avoid needing to deal with certs for local testing/connections
* Due to the above, no password is set

cristianrgreco commented on mkdocs.yml:
```diff
@@ -47,6 +47,7 @@ nav:
       - Cassandra: modules/cassandra.md
       - ChromaDB: modules/chromadb.md
       - Couchbase: modules/couchbase.md
+      - CockroachDB: modules/cockroachdb.md
```
Let's keep this list alphabetical 🙂

leourbina replied:
Haha - I could swear this was alphabetical 🙃

cristianrgreco replied:
Haha I had to double check cos you clearly made effort to keep it alphabetical 😆 Thanks for sorting it

leourbina commented on packages/modules/cockroachdb/src/cockroachdb-container.ts:
```diff
@@ -0,0 +1,71 @@
+
+export class CockroachDbContainer extends GenericContainer {
+  private database = "test";
+  private username = "test";
```
@cristianrgreco do these need to be addressed? I'm following exactly the same pattern as the postgres module.

cristianrgreco replied:
They don't need to be addressed

leourbina commented:
@cristianrgreco Let me know what, if anything, you need from me here to get this merged.

cristianrgreco commented on packages/modules/cockroachdb/src/cockroachdb-container.ts:
```diff
@@ -0,0 +1,71 @@
+    private readonly username: string
+  ) {
+    super(startedTestContainer);
+    this.port = startedTestContainer.getMappedPort(COCKROACH_PORT);
```
This needs to be calculated every time `getPort()` is called because the container can be restarted and get given a different port. E.g:

```js
const container = await new CockroachDbContainer().start();
const port = container.getFirstMappedPort();
await container.restart();
port !== container.getFirstMappedPort()
```

leourbina replied:
Also, worth pointing out that currently this is not working for the postgres container. I'll make a follow up PR to fix it.

leourbina commented:
NP! Just addressed the latest comment and added the test. I had tweak the wait condition to be more robust and ensure that it would work across restarts. Thanks for the thorough review.

cristianrgreco commented on packages/modules/cockroachdb/src/cockroachdb-container.ts:
```diff
@@ -0,0 +1,79 @@
+  }
+
+  public getPort(): number {
+    this.port = this.startedTestContainer.getMappedPort(COCKROACH_PORT);
```
Could just `return this.startedTestContainer.getMappedPort(COCKROACH_PORT);` and remove `this.port`

leourbina replied:
haha good point 🤦

cristianrgreco commented on packages/modules/cockroachdb/src/cockroachdb-container.test.ts:
```diff
@@ -0,0 +1,110 @@
+  });
+  // }
+
+  // workWithRestartedContainer {
```
This is for documentation examples only, we don't need it here

leourbina replied:
Removing

leourbina commented:
@cristianrgreco ok, I think it should be good now.

leourbina commented:
Urgh... ok I think the healthcheck is not as robust as I actually thought. I'll look into it.

leourbina commented:
Hey @cristianrgreco I'm running into the following issue:

Whenever the container starts for the first time, it will run the init scripts. These need to finish running as they setup the user/db and need to be waited for before running anything against it. The only reliable way that I've figured out to detect this is to wait for the log message as so:

```typescript
    this.withExposedPorts(COCKROACH_PORT)
      .withCommand(["start-single-node", "--insecure"])
      .withWaitStrategy(Wait.forLogMessage(/.*end running init files.*/))
      .withStartupTimeout(120_000);
```

This is effectively the same thing that the postgres container is doing [here](https://github.com/testcontainers/testcontainers-node/blob/main/packages/modules/postgresql/src/postgresql-container.ts#L13). 

However, on restarts the init scripts don't run again. I tried fixing it using a healthcheck (as seen on this PR), but I think it is working on my machine as the timing lines up, but based on the test results it is not reliably waiting for the init scripts.

As it is currently written the restart issue we're seeing on the Cockroachdb container exists too on the postgres container. I'm not sure if there are other better ways to ensure that the container is ready both on the first start as well as after a restart. Any ideas here?

leourbina commented:
@cristianrgreco currently the postgres container does not work across restarts either as it suffers from the same issue I pointed above. Namely, the wait condition does not work across restarts. Any chance we can merge this as is? Or do you think we need to fix both CRDB and Postgres as part of this PR?

cristianrgreco commented:
Hi @leourbina, I've raised a PR for fixing the PG container across restarts: https://github.com/testcontainers/testcontainers-node/pull/914

For the Cockroach container I see in the testcontainers-java implementation that they're calling an endpoint ending with `init_success`, which I guess returns 200 once the init is finished - this sounds like what you need:

```java
waitStrategy.withStrategy(
    Wait.forHttp("/health").forPort(REST_API_PORT).forStatusCode(200).withStartupTimeout(Duration.ofMinutes(1))
);
if (this.isVersionGreaterThanOrEqualTo221) {
    waitStrategy.withStrategy(Wait.forSuccessfulCommand("[ -f ./init_success ] || { exit 1; }"));
}
```

cristianrgreco commented:
I've committed a change similar to the PG container healthcheck. Seems to work locally and it allows us to make the interval more frequent for faster startups. Will see what the pipeline sayd

cristianrgreco commented:
Latest changes should be final. @leourbina is there anything more or is this PR good to go?

cristianrgreco commented:
Thanks @leourbina for your contribution!

leourbina commented:
Thank you @cristianrgreco!
